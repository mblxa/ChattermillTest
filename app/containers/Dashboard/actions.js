import { REVIEW_ACTIONS, THEME_ACTIONS } from './constants';

export const listReviews = (offset = 0, theme_id = undefined) => ({
  type: REVIEW_ACTIONS.LIST.CALL,
  offset,
  theme_id,
});
export const listReviewsSuccess = data => ({
  type: REVIEW_ACTIONS.LIST.SUCCESS,
  data,
});
export const listReviewsFail = error => ({
  type: REVIEW_ACTIONS.LIST.FAIL,
  error,
});

export const listThemes = (offset = 0) => ({
  type: THEME_ACTIONS.LIST.CALL,
  offset,
});
export const listThemesSuccess = data => ({
  type: THEME_ACTIONS.LIST.SUCCESS,
  data,
});
export const listThemesFail = error => ({
  type: THEME_ACTIONS.LIST.FAIL,
  error,
});

export const getTheme = id => ({
  type: THEME_ACTIONS.GET.CALL,
  id,
});
export const getThemeSuccess = data => ({
  type: THEME_ACTIONS.GET.SUCCESS,
  data,
});
export const getThemeFail = error => ({
  type: THEME_ACTIONS.GET.FAIL,
  error,
});
