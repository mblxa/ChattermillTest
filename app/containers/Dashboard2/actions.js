import { REVIEW_ACTIONS, THEME_ACTIONS } from './constants';

export const listReviews = (
  offset = 0,
  theme_id = undefined,
  reset = false,
) => ({
  type: REVIEW_ACTIONS.LIST.CALL,
  offset,
  theme_id,
  reset,
});
export const listReviewsSuccess = (data, reset = false) => ({
  type: REVIEW_ACTIONS.LIST.SUCCESS,
  data,
  reset,
});
export const listReviewsFail = error => ({
  type: REVIEW_ACTIONS.LIST.FAIL,
  error,
});

export const listThemes = (offset = 0) => ({
  type: THEME_ACTIONS.LIST.CALL,
  offset,
});
export const listThemesSuccess = data => ({
  type: THEME_ACTIONS.LIST.SUCCESS,
  data,
});
export const listThemesFail = error => ({
  type: THEME_ACTIONS.LIST.FAIL,
  error,
});

export const getTheme = id => ({
  type: THEME_ACTIONS.GET.CALL,
  id,
});
export const getThemeSuccess = data => ({
  type: THEME_ACTIONS.GET.SUCCESS,
  data,
});
export const getThemeFail = error => ({
  type: THEME_ACTIONS.GET.FAIL,
  error,
});
