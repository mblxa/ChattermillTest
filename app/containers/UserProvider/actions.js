import { USER_ACTIONS } from './constants';

export function authorize(username, password) {
  return {
    type: USER_ACTIONS.AUTH.CALL,
    username,
    password,
  };
}

export function checkToken() {
  return {
    type: USER_ACTIONS.CHECK_TOKEN.CALL,
  };
}

export function logout() {
  return {
    type: USER_ACTIONS.LOGOUT.CALL,
  };
}
