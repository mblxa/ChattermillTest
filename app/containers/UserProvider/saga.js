import { call, put, takeEvery } from 'redux-saga/effects';
import moment from 'moment';
import cookie from 'js-cookie';

import { API_ROOT } from '../../api-config';
import request from '../../utils/request';
import { USER_ACTIONS } from './constants';

function* fetchUser(action) {
  const bodyFormData = new FormData();
  bodyFormData.set('username', action.username);
  bodyFormData.set('password', action.password);
  const requestURL = `${API_ROOT}/login`;
  const options = {
    method: 'post',
    mode: 'cors',
    body: bodyFormData,
    // headers: { 'Content-Type': 'multipart/form-data' },
  };

  try {
    const user = yield call(request, requestURL, options);
    yield put({ type: USER_ACTIONS.AUTH.SUCCESS, user });

    const date = moment(user.expire);
    cookie.set('token', user.token, { expires: date.toDate() });
  } catch (e) {
    yield put({ type: USER_ACTIONS.AUTH.FAIL, message: e.message });
  }
}

export default function* userSaga() {
  yield takeEvery(USER_ACTIONS.AUTH.CALL, fetchUser);
}
