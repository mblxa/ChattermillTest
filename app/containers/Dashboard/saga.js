import { call, put, select, takeEvery } from '@redux-saga/core/effects';
import { API_ROOT } from '../../api-config';
import request from '../../utils/request';
import { makeSelectUserToken } from '../UserProvider/selectors';
import {
  getThemeFail,
  getThemeSuccess,
  listReviewsFail,
  listReviewsSuccess,
  listThemesFail,
  listThemesSuccess,
} from './actions';
import { REVIEW_ACTIONS, THEME_ACTIONS } from './constants';
import { makeSelectThemeById } from './selectors';
import {logout} from "../UserProvider/actions";

function* listReviewsSaga(action) {
  const token = yield select(makeSelectUserToken());
  const requestURL = `${API_ROOT}/api/reviews?offset=${action.offset}${
    action.theme_id ? `&theme_id=${action.theme_id}` : ''
  }`;
  const options = {
    method: 'get',
    mode: 'cors',
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const data = yield call(request, requestURL, options);
    yield put(listReviewsSuccess(data.data));
  } catch (e) {
    yield put(listReviewsFail(e.message));
    if (e.code === 401) yield put(logout());
  }
}

function* listThemesSaga(action) {
  const token = yield select(makeSelectUserToken());
  const requestURL = `${API_ROOT}/api/themes?offset=${action.offset}`;
  const options = {
    method: 'get',
    mode: 'cors',
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const data = yield call(request, requestURL, options);
    yield put(listThemesSuccess(data.data));
  } catch (e) {
    yield put(listThemesFail(e.message));
    if (e.code === 401) yield put(logout());
  }
}

function* getThemeSaga(action) {
  const themeSelector = yield select(makeSelectThemeById());

  if (themeSelector(action.id).id !== 0) {
    return true;
  }

  const token = yield select(makeSelectUserToken());
  const requestURL = `${API_ROOT}/api/themes/${action.id}`;
  const options = {
    method: 'get',
    mode: 'cors',
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const data = yield call(request, requestURL, options);
    yield put(getThemeSuccess(data.data));
  } catch (e) {
    yield put(getThemeFail(e.message));
    if (e.code === 401) yield put(logout());
  }
}

// Individual exports for testing
export default function* dashboardSaga() {
  yield takeEvery(REVIEW_ACTIONS.LIST.CALL, listReviewsSaga);
  yield takeEvery(THEME_ACTIONS.LIST.CALL, listThemesSaga);
  yield takeEvery(THEME_ACTIONS.GET.CALL, getThemeSaga);
  // See example in containers/HomePage/saga.js
}
