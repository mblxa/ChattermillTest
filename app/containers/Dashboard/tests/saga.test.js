/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { put } from 'redux-saga/effects';
import { listReviewsSaga } from '../saga';
import { listReviewsFail, listReviewsSuccess } from '../actions';

const token = 123;
describe('dashboardSaga Saga', () => {
  describe('listReviewsSaga', () => {
    let listGenerator;

    // We have to test twice, once for a successful load and once for an unsuccessful one
    // so we do all the stuff that happens beforehand automatically in the beforeEach
    beforeEach(() => {
      listGenerator = listReviewsSaga({ offset: 0 });

      const selectDescriptor = listGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();

      const selectDescriptorToken = listGenerator.next(token).value;
      expect(selectDescriptorToken).toMatchSnapshot();
    });

    it('should dispatch the listReviewsSuccess action if it requests the data successfully', () => {
      const response = { data: [] };

      const putDescriptor = listGenerator.next(response).value;
      expect(putDescriptor).toEqual(put(listReviewsSuccess(response.data)));
    });

    it('should call the listReviewsFail action if the response errors', () => {
      const response = new Error('Some error');
      const putDescriptor = listGenerator.throw(response).value;
      expect(putDescriptor).toEqual(put(listReviewsFail(response.message)));
    });
  });
});
