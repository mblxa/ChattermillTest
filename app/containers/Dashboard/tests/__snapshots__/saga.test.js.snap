// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboardSaga Saga listReviewsSaga should call the listReviewsFail action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`dashboardSaga Saga listReviewsSaga should call the listReviewsFail action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://chattermill-challenge.com/api/reviews?offset=0",
      Object {
        "headers": Object {
          "Authorization": "Bearer 123",
        },
        "method": "get",
        "mode": "cors",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`dashboardSaga Saga listReviewsSaga should dispatch the listReviewsSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`dashboardSaga Saga listReviewsSaga should dispatch the listReviewsSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://chattermill-challenge.com/api/reviews?offset=0",
      Object {
        "headers": Object {
          "Authorization": "Bearer 123",
        },
        "method": "get",
        "mode": "cors",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
